Basic implementation

nforro opened the pull request.

jpopelka commented on specfile/tags.py:
```diff
@@ -0,0 +1,446 @@
+    """
+
+    def __init__(
+        self, data: Optional[List[Tag]] = None, remainder: List[str] = []
+    ) -> None:
+        super().__init__()
+        if data is not None:
+            self.data = data.copy()
+        self._remainder = remainder.copy()
```
[Don't use `[]` as a default argument value (for `remainder`).](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)

```suggestion
        self, data: Optional[List[Tag]] = None, remainder: Optional[List[str]] = None
    ) -> None:
        super().__init__()
        if data is not None:
            self.data = data.copy()
        self._remainder = remainder.copy() if remainder is not None else []
```

nforro replied:
Thanks, I'll fix this.

jpopelka commented on specfile/specfile.py:
```diff
@@ -0,0 +1,101 @@
+        traceback: Optional[types.TracebackType],
+    ) -> None:
+        if not self.autosave:
+            self.save()
```
Is the condition OK? (should the `not` be there?)

nforro replied:
The point of the condition is not to unnecessarily write to the spec file multiple times (at this point it should have been already saved if a change was made), but perhaps it's better to save no matter what.

nforro commented on specfile/tags.py:
```diff
@@ -0,0 +1,445 @@
+        comments = itertools.takewhile(
+            lambda x: x.startswith("#"), reversed(self._preceding_lines)
+        )
+        return [c[1:] for c in reversed(list(comments))]
```
Because the result is supposed to be a list of consecutive comment lines directly above the tag line, so it is necessary to iterate from the bottom up.

nforro commented on specfile/tags.py:
```diff
@@ -0,0 +1,445 @@
+            raise AttributeError(name)
+
+    def find(self, name: str) -> int:
+        try:
+            return next(
+                iter(
+                    i
+                    for i in range(len(self.data))
+                    if self.data[i].name.capitalize() == name.capitalize()
+                )
+            )
+        except StopIteration:
+            raise ValueError
```
No, it doesn't, let me fix that.

FrNecas commented on specfile/rpm.py:
```diff
@@ -0,0 +1,253 @@
+        Parses the content of a spec file.
+
+        Args:
+            s: String representing the content of a spec file.
```
Is there a reason behing this naming? I'd probably much rather see `content` rather than `s`.

nforro replied:
Ok, I'll change it.

FrNecas commented on specfile/rpm.py:
```diff
@@ -0,0 +1,253 @@
+        Args:
+            s: String representing the content of a spec file.
+            sourcedir: Path to sources and patches.
+            macros: List of macro definitions.
```
Maybe a note that this influences the parsing could be useful? Something like `List of macro definitions which will be applied before parsing`

nforro replied:
I assumed it's obvious, but you are right, it's better to be explicit.

FrNecas commented on specfile/specfile.py:
```diff
@@ -0,0 +1,106 @@
+
+class Specfile:
+    """
+    Class representing spec file.
```
Nitpick: missing article for `spec file` (slighly lower `the` is used but `a` is probably more suitable here)

nforro replied:
I was trying to be consistent with other `packit` projects, namely `ogr`, there are no articles in class docstrings there. But it probably wasn't the best idea, I'll add the missing articles everywhere.

FrNecas commented on specfile/specfile.py:
```diff
@@ -0,0 +1,106 @@
+        path: Path to the spec file.
+        sourcedir: Path to sources and patches.
+        autosave: Whether to automatically save any changes made.
+        macros: List of macro definitions.
```
Suggestion: `extra macro definitions` to stress that this is something else than the macros defined in the specfile.

FrNecas commented on specfile/specfile.py:
```diff
@@ -0,0 +1,106 @@
+        self.path = Path(path)
+        self.sourcedir = Path(sourcedir or self.path.parent)
+        self.autosave = autosave
+        self.macros = macros.copy() if macros is not None else None
```
Wouldn't an empty list `[]` be more convenient for further use than `None` if the argument is `None`?

nforro replied:
I don't think there will be any modification to the macros list before passing it to `RPM.parse()`, but sure, I'll change it to `[]` just in case.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,439 @@
+        self.comments, self._preceding_lines = self._extract_comments(preceding_lines)
+
+    def __repr__(self) -> str:
+        preceding_lines = repr(self._preceding_lines)
```
Hm, this doesn't include comments does it? As far as I understand, the input of constructor `preceding_lines` includes all lines before the tag, comments and possibly something "extra" (maybe macro definition?). As far as I can see, this makes this `repr` behave incorrectly, for example if `preceding_lines=['%define macro macro', '# comment']` then the `_preceding_lines` attribute becomes only `['%define macro macro']` and hence the `repr` magic method doesn't actually return the correct way of constructing the object. Am I missing something?

nforro replied:
Good catch, it used to include comments as well, but I changed that and forgot to account for it in `__repr__()`.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,439 @@
+    @staticmethod
+    def _extract_comments(lines: List[str]) -> Tuple[List[str], List[str]]:
+        comments = itertools.takewhile(lambda x: x.startswith("#"), reversed(lines))
+        rest = itertools.dropwhile(lambda x: x.startswith("#"), reversed(lines))
```
This should also probably consider whitespaces before comments, I suggest using a regex (`'  # comment'` should still be valid, I think). Or are whitespaces stripped somewhere else?

nforro replied:
You are right, I'll fix this.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,439 @@
+from specfile.sections import Section
+
+# valid tag names extracted from lib/rpmtag.h in RPM source
+TAG_NAMES = [
```
This should probably be a set (for more efficient check of presence) since we are only checking for presence or iterating over it where order is not relevant.

nforro replied:
Good point.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,439 @@
+            raise AttributeError(name)
+
+    def __setattr__(self, name: str, value: str) -> None:
+        if name.capitalize().rstrip("0123456789") not in TAG_NAMES:
```
Why is there `not in TAG_NAMES`? Shouldn't it be the other way around? Also a minor thing that this lets through tags such as `Arch0`. The number stripping should probably only be done for Patch and Source tags as far as I am aware.

nforro replied:
No, this ensures that "regular" attributes are processed normally (by calling `__setattr__()` of the parent). That's all. You are right about `Arch0`, but I didn't want to make the condition overly complex. Do you think it's necessary?

nforro replied:
Ok, thanks. The rest of the issues should be addressed.

nforro replied:
> Ok, thanks. The rest of the issues should be addressed.

Actually, I just realized that `Tag.comments` and `Tag.__repr__()` is still not entirely correct, I'll fix it ASAP.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,439 @@
+    """
+
+    def __init__(
+        self, data: Optional[List[Tag]] = None, remainder: Optional[List[str]] = None
```
Maybe some documentation about what `remainder` is would be nice.

nforro replied:
It's an internal attribute, and nobody should be instantiating `Tags` directly, but I'll add some description for it.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,439 @@
+        for line in raw_section:
+            m = next((m for m in (r.match(line) for r in tag_regexes) if m), None)
+            if m:
+                e = next(
```
I find this method a bit hard to read/understand at first sight (took me some time to understand what's going on). Maybe some comments could be helpful here, especially due to the use of short names (which is understandable), maybe a quick comment that we need to parse the tag in both the raw and expanded section.

nforro replied:
Fair enough, I'll add some comments.

FrNecas reviewed (changes requested):
Overall looks good to me, I have just a few questions and remarks.

FrNecas reviewed (approved):
Looks good to me now, thanks for the changes. I will try to setup the CI in the upcoming days

nforro commented:
@FrNecas The comments turned out to be a much bigger deal than expected, so I reworked it substantially. It's not possible to use a simple list of strings, because the amount of whitespace preceding every comment must be preserved (in order to ensure minimal diff when making changes).
I've also fixed a bug in classes derived from `collections.UserList` not reimplementing `copy()`, leading to infinite recursion.

FrNecas commented on tests/unit/test_tags.py:
```diff
@@ -0,0 +1,102 @@
+        ],
+        ["%endif"],
+    )
+    assert tags.reassemble() == [
```
Hm, would it be possible to add some comment lines to this test so that it also tests correct reassembly of comments, possibly with whitespaces?

nforro replied:
Good idea, will do.

nforro replied:
Done.

FrNecas commented on specfile/tags.py:
```diff
@@ -0,0 +1,599 @@
+}
+
+
+class Comment:
```
I guess this could be implemented (and slightly simplified) using a [dataclass](https://docs.python.org/3/library/dataclasses.html) since it basically only stores the prefix and text which would remove the need for explicitly defining repr/str

nforro replied:
> which would remove the need for explicitly defining repr/str

For `__repr__()`, yes, but not for `__str__()`.

I'm not sure, is it worth it?

FrNecas replied:
Good point, it probably wouldn't really make much difference, constructor and `__repr__` could be removed but that's about it and it wouldn't simplify it a lot since it's only 2 attributes. I am fine with keeping it the way it is now

FrNecas reviewed (approved):
The changes to comments seem nice, as far as I can tell, nice work. I don't see any more problems with the PR so we can basically merge, just two minor comments from me that could slightly simplify/improve the code.
